sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -1870,6 +1870,37 @@ __ESIMD_API simd<T, N> atomic_update(T *p, simd_view<Toffset, RegionTy> offsets,
       p, offsets, src1, src0, mask);
 }
 
+/// RAII-style class used to implement "semi-dynamic" SLM allocation.
+/// SLM is allocated in the constructor and released in the destructor, that's
+/// why it is "dynamic", as opposed to fully static allocation style of
+/// 'slm_init'. Actual offset of SLM chunk allocated by the call is calculated
+/// at compile time, that's why it is "semi-". To calculate SLM usage by a
+/// kernel, compiler finds a path in a callgraph with the largest amount of SLM
+/// "locked" by slm_allocator objects live along the paths. slm_init call also
+/// participates in calculating SLM budget. It can be modelled as
+/// \c slm_allocator object declared at the very beginning of a kernel and live
+/// till its the very end.
+/// Only compile-time constant SLM amount is supported for now, it is provided
+/// as a class' template argument.
+///
+/// Since a call graph is used, function pointers and recursion is not
+/// supported.
```

Comment:
Actually, no. Recursion is not a problem, because it is not supported by SYCL. But presence of function pointers should be detected, agree. Can diagnostics be added as a follow-up commit? Tagging @v-klochkov.

Reply:
> But presence of function pointers should be detected, agree.

@petercaday, I added diagnostics in https://github.com/intel/llvm/pull/7759/commits/8d20150b2f754d80ddc65f8fb7b00f04fda577d7

Reply:
Thanks, @kbobrovs!